src/main/java/com/mraof/minestuck/block/redstone/TogglerBlock.java

```diff
@@ -0,0 +1,148 @@
+		if(state.getValue(DISCHARGE))
+			worldIn.playSound(null, pos, SoundEvents.PISTON_EXTEND, SoundCategory.BLOCKS, 0.5F, 1.2F);
+		else
+			worldIn.playSound(null, pos, SoundEvents.PISTON_CONTRACT, SoundCategory.BLOCKS, 0.5F, 1.2F);
+		return ActionResultType.SUCCESS;
+	}
+	
+	@Override
+	public void neighborChanged(BlockState state, World worldIn, BlockPos pos, Block blockIn, BlockPos fromPos, boolean isMoving)
+	{
+		super.neighborChanged(state, worldIn, pos, blockIn, fromPos, isMoving);
+		if(fromPos != pos.relative(state.getValue(FACING)))
+			updatePower(worldIn, pos);
+	}
+	
+	@Override
+	public void onPlace(BlockState state, World worldIn, BlockPos pos, BlockState oldState, boolean isMoving)
+	{
+		super.onPlace(state, worldIn, pos, oldState, isMoving);
+		updatePower(worldIn, pos);
+	}
+	
+	public void updatePower(World worldIn, BlockPos pos)
+	{
+		if(!worldIn.isClientSide)
+		{
+			BlockState state = worldIn.getBlockState(pos);
+			boolean hasPower = false;
+			Direction stateFacing = state.getValue(FACING);
+			
+			for(Direction direction : Direction.values()) //checks for a signal in any direction except the one it is facing
+			{
+				if(direction != stateFacing && worldIn.getSignal(pos.relative(direction), direction) > 0)
+				{
+					hasPower = true;
+					break;
+				}
+			}
+			
+			boolean isPoweredBeforeUpdate = state.getValue(POWERED);
+			
+			if(state.getValue(POWERED) != hasPower)
+				worldIn.setBlockAndUpdate(pos, state.setValue(POWERED, hasPower));
+			else worldIn.sendBlockUpdated(pos, state, state, 2);
+			
+			if(!isPoweredBeforeUpdate && hasPower)
+			{
+				BlockPos facingPos = pos.relative(state.getValue(FACING));
+				BlockState facingState = worldIn.getBlockState(facingPos);
+				
+				if(state.getValue(DISCHARGE))
+				{
+					boolean makeSound = false;
+					
+					//removing redstone power from powered blocks
+					for(Property<?> property : facingState.getProperties())
+					{
+						if(facingState.isSignalSource())
+						{
+							if(property.equals(BlockStateProperties.POWERED) || property.equals(BlockStateProperties.POWER))
+							{
+								if(property.equals(BlockStateProperties.POWERED))
+									worldIn.setBlock(facingPos, facingState.setValue(BlockStateProperties.POWERED, false), Constants.BlockFlags.DEFAULT);
+								if(property.equals(BlockStateProperties.POWER))
+									worldIn.setBlock(facingPos, facingState.setValue(BlockStateProperties.POWER, 0), Constants.BlockFlags.DEFAULT);
+								
+								if(!worldIn.getBlockState(pos.relative(state.getValue(FACING).getOpposite())).getBlock().asItem().is(ItemTags.WOOL)) //wont make a toggle sound if the toggler is "muted" by a wool block
+								{
+									makeSound = true;
+								}
+							}
+						}
+					}
+					
+					if(makeSound)
+					{
+						worldIn.playSound(null, facingPos, SoundEvents.UI_BUTTON_CLICK, SoundCategory.BLOCKS, 0.5F, 0.5F);
+					}
+				} else
+				{
+					for(Property<?> property : facingState.getProperties())
```

Comment:
Rewriting this section to use `facingState.hasProperty()` should make this section simpler.